pkg/autodiscovery/listeners/cloudfoundry.go

```diff
@@ -0,0 +1,243 @@
+// Copyright 2017-2020 Datadog, Inc.
+
+// +build clusterchecks
+
+package listeners
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/autodiscovery/integration"
+	"github.com/DataDog/datadog-agent/pkg/util/cloudfoundry"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+func init() {
+	Register("cloudfoundry-bbs", NewCloudFoundryListener)
+}
+
+const (
+	CfServiceContainerIP = "container-ip"
+)
+
+type CloudFoundryListener struct {
+	sync.RWMutex
+	newService    chan<- Service
+	delService    chan<- Service
+	services      map[string]Service // maps ADIdentifiers to services
+	stop          chan bool
+	refreshTicker *time.Ticker
+	bbsCache      cloudfoundry.BBSCacheI
+}
+
+type CloudFoundryService struct {
+	adIdentifier   cloudfoundry.ADIdentifier
+	containerIPs   map[string]string
+	containerPorts []ContainerPort
+	creationTime   integration.CreationTime
+}
+
+// Make sure CloudFoundryService implements the Service interface
+var _ Service = &CloudFoundryService{}
+
+// NewCloudFoundryListener creates a CloudFoundryListener
+func NewCloudFoundryListener() (ServiceListener, error) {
+	bbsCache, err := cloudfoundry.GetGlobalBBSCache()
+	if err != nil {
+		return nil, err
+	}
+	return &CloudFoundryListener{
+		services:      map[string]Service{},
+		stop:          make(chan bool),
+		refreshTicker: time.NewTicker(10 * time.Second),
+		bbsCache:      bbsCache,
+	}, nil
+}
+
+// Listen periodically refreshes services from global BBS API cache
+func (l *CloudFoundryListener) Listen(newSvc chan<- Service, delSvc chan<- Service) {
+	// setup the I/O channels
+	l.newService = newSvc
+	l.delService = delSvc
+
+	go func() {
+		l.refreshServices(true)
+		for {
+			select {
+			case <-l.stop:
+				l.refreshTicker.Stop()
+				return
+			case <-l.refreshTicker.C:
+				l.refreshServices(false)
+			}
+		}
+	}()
+}
+
+func (l *CloudFoundryListener) refreshServices(firstRun bool) {
+	log.Debug("Refreshing services via CloudFoundryListener")
+	allActualLRPs, desiredLRPs := l.bbsCache.GetAllLRPs()
+
+	// if not found and running, add it
+	// at the end, compare what we saw and what is cached and kill what's not there anymore
+	notSeen := make(map[string]interface{})
+	for i := range l.services {
+		notSeen[i] = nil
+	}
+
+	adIdentifiers := l.getAllADIdentifiers(desiredLRPs, allActualLRPs)
+	for _, id := range adIdentifiers {
+		strID := id.String()
+		if _, found := l.services[strID]; found {
+			// delete is no-op when we try to delete a key that doesn't exist
+			// NOTE: this will remove old versions of services on redeploys because ADIdentifier contains ProcessGUID,
+			//       which changes by redeploying
+			delete(notSeen, strID)
+			continue
+		}
+		svc := l.createService(id, firstRun)
+		// if the container is not in RUNNING state, we can't populate ports, thus createService would return nil
+		if svc != nil {
+			l.newService <- svc
+		}
+	}
+
+	for adID := range notSeen {
+		l.Lock()
```

Comment:
I'm guessing we're locking to protect the `delete` operation, if so, should we also lock the for line 89 ?

Also when would we have multiple goroutine accessing `CloudFoundryListener` at the same time ? The `Listen` function should protect from multiple access, no ?

Reply:
You're right that we should also lock for line 89, I'll fix it. The `Listen` function does protect from multiple access, but the `refreshServices` function is triggered by a ticker, so it might potentially be triggered while a previous run is still in progress... But I guess that means that we should just lock in the beginning of the `refreshServices` method and defer unlocking until the function is over. That will be much cleaner and safer. Does that make sense?

Reply:
Yes, thanks for the details !